### PR TITLE
Make elements of SFlux, SMono covariant

### PR DESCRIPTION
**Pull request by jpork**
solving issue SFlux Element Type Should be Declared as Covariant Type  #54

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@jpork Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/reactor?repositoryId=reactor/reactor-scala-extensions&pullRequestId=55)!

[Click here](https://cla.pivotal.io/sync/reactor?repositoryId=reactor/reactor-scala-extensions&pullRequestId=55) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by sinwe** [comment]
Thank you @jpork, it looks like it still fails on scala 2.11. You could try to run `sbt +test`

**Comment by jpork** [comment]
I'm not sure if Scala-2.11 support is still worth it. Though I made it compile & test clean for 2.11.

**Comment by sinwe** [comment]
> I'm not sure if Scala-2.11 support is still worth it. Though I made it compile & test clean for 2.11.

Yes, I agree. April 2020 statistic shows the 2.11 version account only about 6.5%. I'll see the stats in May 2020 and decide if it's worth the trouble to support it as there are lots of deprecated that remains in the code just to support 2.11.

With regards to the PR, it has passed all the checks except 1, the ci/pivotal-cla. You need to sign the cla (contributor license agreement). This requirement is not from me, but from pivotal.

File: src/main/scala/reactor/core/scala/publisher/PimpMyPublisher.scala [hunk 1]
```diff
@@ -11,17 +11,17 @@ import scala.language.implicitConversions
...
   implicit def jFluxJInt2JFluxInt(jFluxInt: JFlux[Integer]): JFlux[Int] = jFluxInt.map[Int]((i: Integer) => Integer2int(i))
 
   implicit def jMonoJLong2JMonoLong(mono: JMono[JLong]): JMono[Long] = mono.map(Long2long(_: JLong))
 
-  implicit def jFlux2SFlux[T](jFlux: JFlux[T]): SFlux[T] = new ReactiveSFlux[T](jFlux)
-
-  implicit def jMono2SMono[T](jMono: JMono[T]): SMono[T] = new ReactiveSMono[T](jMono)
+//  implicit def jFlux2SFlux[T](jFlux: JFlux[T]): SFlux[T] = new ReactiveSFlux[T](jFlux)
+//
+//  implicit def jMono2SMono[T](jMono: JMono[T]): SMono[T] = new ReactiveSMono[T](jMono)
```
Comment (rs017991) [inline src/main/scala/reactor/core/scala/publisher/PimpMyPublisher.scala:26]: are these no longer possible?

Comment (sinwe) [reply src/main/scala/reactor/core/scala/publisher/PimpMyPublisher.scala:26]: Personally I don't think this is even needed anymore. But I could be wrong.

Comment (sinwe) [reply src/main/scala/reactor/core/scala/publisher/PimpMyPublisher.scala:26]: Personally I don't think this is even needed anymore. But I could be wrong. `PimpMyPublisher` is open for public, so developer could use it. When I wrote this, it was meant for internal usage and for application to use `.asJava` and `.asScala` explicitly to make the code more readable.

File: src/main/scala/reactor/core/scala/publisher/SFlux.scala [hunk 2]
```diff
@@ -654,7 +685,13 @@ trait SFlux[T] extends SFluxLike[T, SFlux] with MapablePublisher[T] with ScalaCo
     */
   final def publishNext(): SMono[T] = SMono.fromPublisher(coreFlux.publishNext())
 
-  final def reduce(aggregator: (T, T) => T): SMono[T] = coreFlux.reduce(aggregator).asScala
+  final def reduce[U >: T](aggregator: (U, U) => U): SMono[U] = {
+    //def r[P <: T](v1: U, v2: U): P = aggregator(v1, v2).asInstanceOf[P]
```
Comment (sinwe) [inline src/main/scala/reactor/core/scala/publisher/SFlux.scala:689]: Can this commented line be removed?

File: src/main/scala/reactor/core/scala/publisher/SFluxLike.scala [hunk 3]
```diff
@@ -41,7 +40,7 @@ trait SFluxLike[T, Self[U] <: SFluxLike[U, Self]] extends ScalaConverters {
 
   final def foldLeft[R](initial: R)(binaryOps: (R, T) => R): SMono[R] = reduce[R](initial)(binaryOps)
 
-  final def head: SMono[T] = take(1).as(SMono.fromPublisher)
+  final def head: SMono[T] = coreFlux.next.asScala
```
Comment (sinwe) [inline src/main/scala/reactor/core/scala/publisher/SFluxLike.scala:43]: Nice. I didn't realize there is `.next` operator.

File: src/main/scala/reactor/core/scala/publisher/PimpMyPublisher.scala (hunk 1 above)
Comment (sinwe) [reply src/main/scala/reactor/core/scala/publisher/PimpMyPublisher.scala:26]: @jpork Could this commented line be removed?

Comment (jpork) [reply src/main/scala/reactor/core/scala/publisher/PimpMyPublisher.scala:26]: From my point of view I wouldn't provide implicit conversion between SFlux & Co and JFlux & Co. This is to dangerous and adding .asJava or .asScala as conversion code is not to much and better readable. 
Scala collections do it the same way. This conversion provoked a recursive call in SFlux which got on my nerves for quite a while. 
So I would remove the whole PimpMyPublisher thing.  But this is another story.

File: src/main/scala/reactor/core/scala/publisher/SFlux.scala (hunk 2 above)
Comment (jpork) [reply src/main/scala/reactor/core/scala/publisher/SFlux.scala:689]: yup. sorry leaving that.

**Comment by pivotal-issuemaster** [comment]
@jpork Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/reactor?repositoryId=reactor/reactor-scala-extensions&pullRequestId=55)!

**Comment by jpork** [comment]
removed the out commented code and signed the cla.

File: src/main/scala/reactor/core/scala/publisher/PimpMyPublisher.scala (hunk 1 above)
Comment (sinwe) [reply src/main/scala/reactor/core/scala/publisher/PimpMyPublisher.scala]: > This conversion provoked a recursive call in SFlux which got on my nerves for quite a while.
> So I would remove the whole PimpMyPublisher thing. But this is another story.

Totally agree with this. Sometimes i also got recursive call to `SFlux` itself while the intention is to call `JFlux`.

**Pull request merged**